maintainer_guide.md

```diff
@@ -0,0 +1,80 @@
+## Torchvision maintainers guide
+
+This document aims at documenting common questions and workflows that arise when
+developing and maintaining torchvision. It does not serve as an absolute ground
+truth, but rather as a handy reference guide.
+
+### What is public?
+
+* In the torchvision folder:
+    * For Python, the [usual
+      convention](https://www.python.org/dev/peps/pep-0008/#public-and-internal-interfaces)
+      applies: anything that has a leading underscore in its name or in its path
+      is private. The rest is public, even if it’s not properly documented, and
+      even if it’s not exposed in an `__init__` file. While we discourage use of
```

Comment:
> The rest is public, even if it’s not properly documented, and even if it’s not exposed in an `__init__` file. 

The use of word "public" in this statement feels to me giving out impression that these interfaces are BC-guaranteed.

If the goal of this section is to encourage developers and maintainers to think twice about the backward compatibility of such interfaces, I think we can focus on reality more.

```
Interfaces excluded from documentation and `__init__` attributes are not guaranteed for backward compatibility. (For reasons such as residuals from early stage of the project) Therefore, users should not rely on them, (or users find them useful are encourage to file an issue to make them into a public interface) however, maintainers must still consider backwards compatibility when making changes to such potentially in-use code.
```

Reply:
> The use of word "public" in this statement feels to me giving out impression that these interfaces are BC-guaranteed.

Yes, unfortunately they are. I believe this is what we all agreed on during our internal discussions.

Reply:
What I remember from our discussion was that we want to communicate that maintainers should consider and follow appropriate steps for BC-breaking changes to these kinds of APIs, but that we should avoid implying a particular guarantee to users.

What a user should expect is a different question we could address elsewhere, but we can write our maintainer instructions without additional implications for users. I don't think we have any absolute "guarantees" of backwards compatibility, just that we expect maintainers to follow this process when breaking BC.

Elsewhere we can communicate different standards for backwards compatibility, which might consider whether the API is documented. Our actual communicated-to-users level of BC commitment might include differences for major/minor releases vs nightlies/master, documented vs undocumented APIs, etc.

Reply:
I agree with your comment @nairbv 

Regarding of the use of the term "public", I'm not sure why this is so controversial: we're simply following pep8 here.

To clarify my point above:


> > The use of word "public" in this statement feels to me giving out impression that these interfaces are BC-guaranteed.
> 
> Yes, unfortunately they are. I believe this is what we all agreed on during our internal discussions.


An interface  can be either public or private. From pep8, an interface is private if and only if it has a leading underscore, and it's public otherwise. Private interfaces can be broken without BC considerations. Public interfaces cannot be broken without BC considerations.

What those *considerations* are is **not** the purpose of this section. They are instead (unfortunately only partially described) below.

As an important datapoint that I already mentionned, scikit-learn and numpy both consider that all public interfaces follow the same deprecation cycles, whether those were initially intended to be effectively public or not.